app/javascript/components/student/Editor.tsx

```diff
@@ -0,0 +1,45 @@
+import React, { useState } from 'react'
+import { CodeEditor } from './editor/CodeEditor'
+import { TestRunSummary } from './editor/TestRunSummary'
+import fetch from 'isomorphic-fetch'
+
+export type Submission = {
+  testsStatus: TestRunStatus
+  uuid: string
+}
+
+export enum TestRunStatus {
+  PASS = 'pass',
+  FAIL = 'fail',
+  ERROR = 'error',
+  PENDING = 'pending',
+  OPS_ERROR = 'ops_error',
+}
+
+export function Editor({ endpoint }: { endpoint: string }) {
+  const [submission, setSubmission] = useState<Submission>()
+
+  function submit(code: string) {
+    setSubmission(undefined)
+
+    fetch(endpoint, {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify({ files: { file: code } }),
+    })
+      .then((response) => response.json())
```

Comment:
Consider:

1. First check if the request is in the 1xx or 2xx range:

```typescript
if (!response.ok) {
  throw response
}

// More here
```

Some people frown upon throwing non promise values, but I don't. It's a perfectly valid way to separate error cases from the happy path, without wrapping it in a custom error, which you can do. I'll leave that up to you.

2. Check the response content type to determine if you can JSONfy this:

```typescript
const contentType = response.headers.get('content-type')

if (!contentType.includes('+json') && !contentType.includes('application/json')) {
  // Unexpected content type. Optionally wrap
  // in a custom error, for example "Unexpected
  // 'content-type'. Expected ..., actual ...
  throw response
}

return response.json()
```

3. Handle the error _response_
```typescript
.catch((errorOrResponse) => {

  // Probably want to invert this statement, but
  // I am on my phone and was too much work.
  // Inverting leads to early return and less
  // nested code.
  if (errorOrResponse instanceof Response) {

    // Convert response to error. I usually check
    // the 'content-type' here and determine
    // how to deserialize the error. I then have
    // custom error classes per type:
    //
    // - StructuredError where I try to extract the
    //   error message from the JSON.
    // - TextError where I just forward the text in
    //   the response as the error message
    // - Client side 415 meaning Unsupported 
    //   MediaType. 
    //
    // I'd prefix the error message with response
    // status code and maybe statusText.
  }

   return Promise.reject(errorOrResponse)
})
```

You can look at my npm package `fetch-media` for inspiration. I recommend for now not using that and extracting this in a "roll your own" utility, because we need to explain to people how this works and otherwise it will be "abstracted away" in a package.